apis/condition_status_type.go

```diff
@@ -0,0 +1,121 @@
+	// Conditions communicates information about ongoing/complete
+	// reconciliation processes that bring the "spec" inline with the observed
+	// state of the world.
+	// +optional
+	Conditions []Condition `json:"conditions,omitempty"`
+}
+
+// ConditionStatus defines a readiness condition for a Knative resource.
+// See: https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#typical-status-properties
+type Condition struct {
+	// Type of condition.
+	// +required
+	Type ConditionType `json:"type" description:"type of status condition"`
+
+	// Status of the condition, one of True, False, Unknown.
+	// +required
+	Status corev1.ConditionStatus `json:"status" description:"status of the condition, one of True, False, Unknown"`
+
+	// LastTransitionTime is the last time the condition transitioned from one status to another.
+	// We use VolatileTime in place of metav1.Time to exclude this from creating equality.Semantic
+	// differences (all other things held constant).
+	// +optional
+	LastTransitionTime VolatileTime `json:"lastTransitionTime,omitempty" description:"last time the condition transit from one status to another"`
+
+	// The reason for the condition's last transition.
+	// +optional
+	Reason string `json:"reason,omitempty" description:"one-word CamelCase reason for the condition's last transition"`
+
+	// A human readable message indicating details about the transition.
+	// +optional
+	Message string `json:"message,omitempty" description:"human-readable message indicating details about last transition"`
+}
+
+// IsReady looks at the conditions on the ConditionStatus.
+// Returns true if condition[Ready].Status is true.
+func (cs *ConditionStatus) IsReady() bool {
+	if c := cs.GetCondition(ConditionReady); c != nil {
+		return c.Status == corev1.ConditionTrue
+	}
+	return false
+}
+
+// GetCondition finds and returns the Condition that matches the ConditionType previously set on ConditionStatus.
+func (cs *ConditionStatus) GetCondition(t ConditionType) *Condition {
+	for _, cond := range cs.Conditions {
+		if cond.Type == t {
+			return &cond
+		}
+	}
+	return nil
+}
+
+// SetCondition sets or updates the Condition on ConditionStatus for Condition.Type.
+func (cs *ConditionStatus) SetCondition(new *Condition) {
+	if new == nil {
+		return
+	}
+	t := new.Type
+	var conditions []Condition
+	for _, cond := range cs.Conditions {
+		if cond.Type != t {
+			conditions = append(conditions, cond)
+		} else {
+			// If we'd only update the LastTransitionTime, then return.
+			new.LastTransitionTime = cond.LastTransitionTime
+			if reflect.DeepEqual(new, &cond) {
+				return
+			}
+		}
+	}
+	new.LastTransitionTime = VolatileTime{metav1.NewTime(time.Now())}
+	conditions = append(conditions, *new)
+	sort.Slice(conditions, func(i, j int) bool { return conditions[i].Type < conditions[j].Type })
+	cs.Conditions = conditions
+}
+
+// InitializeConditions updates the Ready Condition to unknown if not set.
```

Comment:
This should iterate over a set of conditions which are present on the resource. I'm not sure how you define the set of conditions across all resources, though.

https://github.com/knative/serving/blob/master/pkg/apis/serving/v1alpha1/revision_types.go#L310